shared/src/main/scala/mlscript/helpers.scala

```diff
@@ -80,10 +128,166 @@ abstract class TypeImpl extends Located { self: Type =>
+        (SourceCode(s"type res") ++
+          SourceCode.paramList(ctx.newTypeParams) ++
+          SourceCode.equalSign ++
+          tsType
+        ) +
+        SourceCode("// end ts")
+    }
+  }
+
+  /**
+    * Converts an mlscript type to source code representation of a ts type. It also
+    * keep tracks of extra type variables and type parameters defined through the context.
+    *
+    * Takes a context to maintain state.
+    * @param ctx
+    * @return
+    */
+  def toTsType(implicit ctx: ToTsTypegenContext = ToTsTypegenContext.empty): SourceCode = {
+    this match {
+      // these types do not mutate typegen context
+      case Union(TypeName("true"), TypeName("false")) | Union(TypeName("false"), TypeName("true")) => SourceCode("boolean")
+      case Union(lhs, rhs) => SourceCode.sepBy(List(lhs.toTsType, rhs.toTsType), SourceCode.separator)
+      case Inter(lhs, rhs) => SourceCode.sepBy(List(lhs.toTsType, rhs.toTsType), SourceCode.ampersand)
+      case Record(fields) => SourceCode.recordWithEntries(fields.map(entry => (SourceCode(entry._1.name), entry._2.toTsType)))
+      // unwrap extra parenthesis when tuple has single element
+      case Tuple(fields) => if (fields.length === 1) {
+        fields(0)._2.toTsType
+      } else {
+        SourceCode.horizontalArray(fields.map(field => field._2.toTsType))
+      }
+      case Top => SourceCode("unknown")
+      case Bot => SourceCode("never")
+      case TypeName(name) => SourceCode(name)
+      case Literal(IntLit(n)) => SourceCode(n.toString + (if (JSBackend isSafeInteger n) "" else "n"))
+      case Literal(DecLit(n)) => SourceCode(n.toString)
+      case Literal(StrLit(s)) => SourceCode(JSLit.makeStringLiteral(s))
+      
+      // these types may mutate typegen context by argCounter, or
+      // by creating new type aliases
+      case Function(lhs, rhs) =>
+          val currArgCount = ctx.argCounter;
+          val currPolarity = ctx.polarity;
+          
+          ctx.argCounter += 1;
+          ctx.flipPolarity
+          val lhsTypeSource = lhs.toTsType;
+          ctx.polarity = currPolarity;
+          
+          val rhsTypeSource = rhs.toTsType;
+          
+          (SourceCode(s"arg${currArgCount}") ++ SourceCode.colon ++ lhsTypeSource).parenthesized ++
+          SourceCode.fatArrow ++
+          rhsTypeSource
+      // a recursive type is wrapped in a self referencing Recursion type
+      // this wrapped form is used only inside it's body
+      case Recursive(uv, body) =>
+        val uvTsType = uv.toTsType
+        
+        // use modified context for converting body type. This will replace
+        // all use of uv type var with it's wrapped type.
+        ctx.newTypeAlias = (SourceCode("type ") ++ uvTsType ++
+          SourceCode.equalSign ++ body.toTsType) +: ctx.newTypeAlias
+          
+        // recursive type is no longer a type variable it's an alias instead
+        ctx.existingTypeVars = ctx.existingTypeVars.removed(uv)
```

Comment:
What does this line achieve? What if there happens to be the same recursive variable used twice, as in `(... as 'a) -> (... as 'a)`? This logic is very fishy.